File: NOVEMBER-2025/21-11-2025/p1930.rb
# @param {String} s
# @return {Integer}
def count_palindromic_subsequence(s)
  n = s.length
  first = Array.new(26, -1)
  last = Array.new(26, -1)
  bytes = s.bytes

  (0...n).each do |i|
    c = bytes[i] - 97
    first[c] = i if first[c] == -1
    last[c] = i
  end

  ans = 0
  (0...26).each do |c|
    if first[c] != -1 && last[c] - first[c] > 1
      mask = 0
      ((first[c] + 1)...last[c]).each do |i|
        mask |= 1 << (bytes[i] - 97)
      end
      ans += mask.to_s(2).count('1')
    end
  end

  ans
end

File: NOVEMBER-2025/21-11-2025/p1930.php
class Solution {

    /**
     * @param String $s
     * @return Integer
     */
    function countPalindromicSubsequence($s) {
        $n = strlen($s);
        $first = array_fill(0, 26, -1);
        $last = array_fill(0, 26, -1);

        for ($i = 0; $i < $n; $i++) {
            $c = ord($s[$i]) - 97;
            if ($first[$c] == -1) $first[$c] = $i;
            $last[$c] = $i;
        }

        $ans = 0;

        for ($c = 0; $c < 26; $c++) {
            if ($first[$c] != -1 && $last[$c] - $first[$c] > 1) {
                $mask = 0;
                for ($i = $first[$c] + 1; $i < $last[$c]; $i++) {
                    $mask |= 1 << (ord($s[$i]) - 97);
                }
                $ans += substr_count(decbin($mask), '1');
            }
        }

        return $ans;
    }
}